docs/proposals/advanced-nginx-extensions.md

```diff
@@ -15,14 +15,1156 @@ NGINX use cases. To allow them to implement those use cases, we need to bring a
+> contexts of the stream module.
+
+SnippetsFilter supports `http`, `server`, and `location` contexts of the stream module.
+
+> TLSRoute, TCPRoute, and UPDRoute don't support filters. As a result, SnippetsFilter doesn't support stream-related
+> contexts.
+>
+> Snippets for location might share the same problem as mentioned in the issue https://github.com/nginxinc/nginx-gateway-fabric/issues/207,
+> depending on the NGINX directives being used in the snippets. This proposal doesn't address the problem but
+> anticipates the solution to https://github.com/nginxinc/nginx-gateway-fabric/issues/2079 will also solve the problem
+> for Snippets.
+
+#### Examples
+
+Below are a few examples of using snippets to bring unsupported NGINX configuration into NGF.
+
+##### Rate-limiting SnippetPolicy
+
+We use NGINX [limit req module](https://nginx.org/en/docs/http/ngx_http_limit_req_module.html) to configure rate
+limiting.
+
+```yaml
+apiVersion: gateway.nginx.org/v1alpha1
+kind: SnippetsPolicy
+metadata:
+  name: rate-limit
+spec:
+  targetRefs:
+  - group: gateway.networking.k8s.io
+    kind: HTTPRoute
+    name: cafe-route
+  snippets:
+  - context: http
+    value: limit_req_zone $binary_remote_addr zone=myzone:10m rate=1r/s;
+  - context: http.server.location
+    value: limit_req zone=myzone burst=3;
+```
+
+As a result, NGF will insert those snippets into the generated configuration for the referenced HTTPRoute:
+
+```text
+# this is http context
+
+limit_req_zone $binary_remote_addr zone=myzone:10m rate=1r/s;
+. . .
+
+server {
+  . . .
+
+  location /coffee {
+    . . .
+    limit_req zone=myzone burst=3;
+    . . .
+  }
+
+  location /tea {
+    . . .
+    limit_req zone=myzone burst=3;
+    . . .
+  }
+```
+
+##### Proxy Buffering SnippetPolicy
+
+We configure [proxy_buffering](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_buffering) directive
+to disable buffering.
+
+```yaml
+apiVersion: gateway.nginx.org/v1alpha1
+kind: SnippetsPolicy
+metadata:
+  name: buffering
+spec:
+  targetRefs:
+  - group: gateway.networking.k8s.io
+    kind: HTTPRoute
+    name: cafe-route
+  snippets:
+  - context: http.server.location
+    value: proxy_buffering off;
+```
+
+As a result, NGF will insert the provided config into the generated locations for the cafe-route HTTPRoute.
+
+##### Access Control SnippetPolicy
+
+We use NGINX [access module](https://nginx.org/en/docs/http/ngx_http_access_module.html) to configure access based
+on client IPs.
+
+```yaml
+apiVersion: gateway.nginx.org/v1alpha1
+kind: SnippetsPolicy
+metadata:
+  name: access-control
+spec:
+  targetRefs:
+  - group: gateway.networking.k8s.io
+    kind: Gateway
+    name: cafe
+  snippets:
+  - context: http.server
+    value: |
+      allow 10.0.0.0/8;
+      deny all;
+```
+
+As a result, NGF will insert the provided NGINX config into the server context of all generated HTTP servers for
+the Gateway cafe.
+
+##### Access Control SnippetsFilter
```

Comment:
So in theory this could either be attached as a Policy to an HTTPRoute, or as a filter to an HTTPRoute, and the behavior would be the same?

Reply:
yep. NGINX config will look the same for the happy path.

However, the biggest difference will be in what happens if Snippets resource is invalid or missing:
- With SnippetsPolicy. NGF will ignore such a policy, and NGINX will allow traffic from any clients to the backends.
- With SnippetsFilter. NGF will notice that the referenced Filter in the HTTPRoute is invalid/missing, and it will add `return 500` to the NGINX config for the affected locations, so that NGINX will return a 500 response to clients (in contrast with allowing traffic from any clients).